### PR TITLE
feat: add Balancer as trade provider

### PR DESCRIPTION
**Pull request by guanbinrui**
## Tasks

+ [x] Add balancer logo and provider links.
+ [x] Add balancer constants (GAS_PRICE, MAX_HOPS, CONTRACT_ADDRESS etc).
+ [x] Add contract ABI and hook.
+ [x] Add trade hooks.
    + [x] `useTrade`
    + [x] `useTradeCallback`
    + [x] `useTradeComputed`
+ [x] Fetch swaps from SOR.
+ [x] Add a routes viewer.
+ [ ] Add a pool viewer (maybe do it in the future).

### Mainnet network (could be skipped)

+ [x] Compare price and fee with the original front-end app.
+ [ ] Test Ether -> ERC20
+ [ ] Test ERC20 -> Ether
+ [ ] Test ERC20 -> ERC20

### Kovan network

+ [ ] Compare price and fee with the original front-end app.
+ [ ] Test Ether -> ERC20
+ [ ] Test ERC20 -> Ether
+ [ ] Test ERC20 -> ERC20

## Screenshots

test

## Learn more

+ [Balancer Exchange](https://balancer.exchange/)
+ [smart order router (SOR)](https://docs.balancer.finance/smart-contracts/sor/development)
+ [exchange-proxy](https://docs.balancer.finance/smart-contracts/exchange-proxy)

### CODE REVIEW

File: packages/maskbook/src/resources/BalancerIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { SvgIconProps, SvgIcon } from '@material-ui/core'
+
+const svg = (
+    <svg width="24" height="24" viewBox="0 0 258 258" version="1.1" xmlns="http://www.w3.org/2000/svg">
```
Comment (septs) [inline packages/maskbook/src/resources/BalancerIcon.tsx:4]: https://aka.pw/svgomg

File: packages/maskbook/src/plugins/Trader/constants/trader.ts [hunk 2]
```diff
@@ -46,6 +46,28 @@ export const TRADE_CONSTANTS = {
         [ChainId.Kovan]: '',
         [ChainId.Gorli]: '',
     },
+    BALANCER_ETH_ADDRESS: {
+        [ChainId.Mainnet]: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/constants/trader.ts:50]: ?

Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/constants/trader.ts:50]: This is an imaginary address (just like 0x0....0). In addition, It's checksummed.

File: packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            alignItems: 'center',
+        },
+        name: {
+            fontSize: 12,
+            marginLeft: theme.spacing(1),
+        },
+    }),
+)
+
+export interface TradeRouteProps extends withClasses<KeysInferFromUseStyles<typeof useStyles>> {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx:77]: Please stop using `KeysInferFromUseStyles` in new code. It's deprecated.

File: packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                </TableHead>
+                <TableBody>
+                    {props.trade.trade_.routes.map((route, i) => (
+                        <TableRow>
+                            <TableCell className={classes.cell} align="left" key={i}>
+                                <Breadcrumbs
+                                    classes={{ ol: classes.list, li: classes.item }}
+                                    separator={<NavigateNextIcon fontSize="small" />}>
+                                    {route.hops.map((hop) => {
+                                        return (
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx:108]: code goes too deep. would you like to split a new comp at this point?

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx (hunk 3 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx:77]: Removed!

File: packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx (hunk 4 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx:108]: Done!

File: packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            alignItems: 'center',
+        },
+        name: {
+            fontSize: 12,
+            marginLeft: theme.spacing(1),
+        },
+    }),
+)
+
+export interface TradeRouteProps extends withClasses<never> {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/UI/balancer/TradeRoute.tsx:69]:
```suggestion
export interface TradeRouteProps {
```

**Review by Jack-Works — COMMENTED** [review]
generally lgtm on code style

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**